src/promptflow-tools/tests/test_open_source_llm.py

```diff
@@ -0,0 +1,166 @@
+
+    @pytest.mark.skip_if_no_key("gpt2_custom_connection")
+    def test_open_source_llm_con_url_chat(self, gpt2_custom_connection):
+        del gpt2_custom_connection.configs['endpoint_url']
+        with pytest.raises(OpenSourceLLMKeyValidationError) as exc_info:
+            os = OpenSourceLLM(gpt2_custom_connection)
+            os.call(self.chat_prompt, API.CHAT)
+        assert exc_info.value.message == """Required key `endpoint_url` not found in given custom connection.
+Required keys are: endpoint_url,model_family."""
+
+    @pytest.mark.skip_if_no_key("gpt2_custom_connection")
+    def test_open_source_llm_con_key_chat(self, gpt2_custom_connection):
+        del gpt2_custom_connection.secrets['endpoint_api_key']
+        with pytest.raises(OpenSourceLLMKeyValidationError) as exc_info:
+            os = OpenSourceLLM(gpt2_custom_connection)
+            os.call(self.chat_prompt, API.CHAT)
+        assert exc_info.value.message == ("Required secret key `endpoint_api_key` "
+            + """not found in given custom connection.
+Required keys are: endpoint_api_key.""")
+
+    @pytest.mark.skip_if_no_key("gpt2_custom_connection")
+    def test_open_source_llm_con_model_chat(self, gpt2_custom_connection):
+        del gpt2_custom_connection.configs['model_family']
+        with pytest.raises(OpenSourceLLMKeyValidationError) as exc_info:
+            os = OpenSourceLLM(gpt2_custom_connection)
+            os.call(self.completion_prompt, API.COMPLETION)
+        assert exc_info.value.message == """Required key `model_family` not found in given custom connection.
+Required keys are: endpoint_url,model_family."""
+
+    def test_open_source_llm_escape_chat(self):
+        danger = r"The quick \brown fox\tjumped\\over \the \\boy\r\n"
+        out_of_danger = ContentFormatterBase.escape_special_characters(danger)
+        assert out_of_danger == "The quick \\brown fox\\tjumped\\\\over \\the \\\\boy\\r\\n"
+
+    def test_open_source_llm_llama_parse_chat_with_chat(self):
+        LlamaContentFormatter.parse_chat(self.chat_prompt)
+
+    def test_open_source_llm_llama_parse_multi_turn(self):
+        multi_turn_chat = """user:
+You are a AI which helps Customers answer questions.
+
+What is the best movie of all time?
+
+assistant:
+Mobius, which starred Jared Leto
+
+user:
+Why was that the greatest movie of all time?
+"""
+        LlamaContentFormatter.parse_chat(multi_turn_chat)
+
+    def test_open_source_llm_llama_parse_system_not_accepted(self):
+        bad_chat_prompt = """system:
+You are a AI which helps Customers answer questions.
+
+user:
+""" + self.completion_prompt
+        with pytest.raises(OpenSourceLLMUserError) as exc_info:
+            LlamaContentFormatter.parse_chat(bad_chat_prompt)
+        assert exc_info.value.message == ("The Chat API requires a specific format for prompt definition,"
+            + " and the prompt should include separate lines as role delimiters: 'assistant:\\n','user:\\n'."
```

Comment:
Please assert the error code hierarchy for all tests about your new error like the code in https://github.com/microsoft/promptflow/blob/main/src/promptflow-tools/tests/test_handle_openai_error.py

For example,  "assert exc_info.value.error_codes == error_codes.split("/")"